File: resources/ios/Sources/ConsentManager.swift
import Foundation
import UserMessagingPlatform

/**
 * Surrounds the UMP ConsentInformation singleton: debug-settings construction
 * (so the consent form can be forced during testing) and the consentStatus ->
 * PHP-string mapping aligned with ConsentChanged::STATUS_* on the Laravel side.
 *
 * NOTE: GoogleUserMessagingPlatform ~> 2.7 exposes Swift-refined, de-prefixed
 * type names (ConsentInformation / ConsentForm / RequestParameters /
 * DebugSettings / DebugGeography / ConsentStatus), matching how this repo
 * already uses the de-prefixed Google Mobile Ads names (BannerView, Request,
 * AppOpenAd). If the pinned pod resolves only the UMP-prefixed ObjC names,
 * switch these to UMPConsentInformation.sharedInstance etc.
 */
enum ConsentManager {
    static var info: ConsentInformation { ConsentInformation.shared }

    static func requestParameters() -> RequestParameters {
        let params = RequestParameters()
        if let debug = debugSettings() {
            params.debugSettings = debug
        }

        return params
    }

    /**
     * Builds DebugSettings from env so the consent form's geography can be
     * forced during testing. ADMOB_UMP_DEBUG_GEOGRAPHY is one of
     * EEA / NOT_EEA / DISABLED. (Test devices are managed in the AdMob console,
     * not here; or simply use a VPN to a EEA region to exercise the real form.)
     */
    private static func debugSettings() -> DebugSettings? {
        let env = ProcessInfo.processInfo.environment
        let geography = env["ADMOB_UMP_DEBUG_GEOGRAPHY"]?.trimmingCharacters(in: .whitespaces).uppercased()

        if geography == nil || geography == "" || geography == "DISABLED" {
            return nil
        }

        let settings = DebugSettings()
        switch geography {
        case "EEA": settings.geography = .EEA
        case "NOT_EEA", "NON_EEA": settings.geography = .notEEA
        default: settings.geography = .disabled
        }

        return settings
    }

    /**
     * Maps ConsentInformation.consentStatus to the PHP string constants in
     * BlessedZulu\NativePhpAdmob\Events\ConsentChanged.
     */
    static func statusString() -> String {
        switch info.consentStatus {
        case .notRequired: return "not_required"
        case .required: return "required"
        case .obtained: return "obtained"
        default: return "unknown"
        }
    }

    static func isFormRequired() -> Bool {
        return info.consentStatus == .required
    }
}

File: resources/android/src/ConsentManager.kt
package com.blessedzulu.nativephp.admob

import android.content.Context
import androidx.fragment.app.FragmentActivity
import com.google.android.ump.ConsentDebugSettings
import com.google.android.ump.ConsentInformation
import com.google.android.ump.ConsentRequestParameters
import com.google.android.ump.UserMessagingPlatform

/**
 * Holds the process-wide UMP ConsentInformation singleton and the helpers that
 * surround it: debug-settings construction (so the consent form can be forced
 * on a non-EEA test device) and the int-status -> PHP-string mapping that keeps
 * the native layer aligned with ConsentChanged::STATUS_* on the Laravel side.
 *
 * Mirrors the registry pattern used by the ad formats - a single object that
 * owns one piece of SDK state for the whole app.
 */
object ConsentManager {
    @Volatile
    private var cached: ConsentInformation? = null

    @Synchronized
    fun info(context: Context): ConsentInformation =
        cached ?: UserMessagingPlatform.getConsentInformation(context.applicationContext)
            .also { cached = it }

    fun requestParameters(context: Context): ConsentRequestParameters {
        val builder = ConsentRequestParameters.Builder()
        buildDebugSettings(context)?.let { builder.setConsentDebugSettings(it) }

        return builder.build()
    }

    /**
     * Builds ConsentDebugSettings from env so the form can be triggered during
     * testing. ADMOB_UMP_DEBUG_GEOGRAPHY is one of EEA / NOT_EEA / DISABLED.
     * ADMOB_TEST_DEVICES is a comma-separated list of UMP hashed device IDs
     * (the value the UMP SDK prints to logcat on the first info-update on an
     * unconfigured device - NOT the MobileAds test-device ID).
     */
    private fun buildDebugSettings(context: Context): ConsentDebugSettings? {
        val geography = System.getenv("ADMOB_UMP_DEBUG_GEOGRAPHY")?.trim()?.uppercase()
        val devices = System.getenv("ADMOB_TEST_DEVICES")
            ?.split(",")
            ?.map { it.trim() }
            ?.filter { it.isNotBlank() }
            ?: emptyList()

        if ((geography.isNullOrBlank() || geography == "DISABLED") && devices.isEmpty()) {
            return null
        }

        val builder = ConsentDebugSettings.Builder(context)
        when (geography) {
            "EEA" -> builder.setDebugGeography(ConsentDebugSettings.DebugGeography.DEBUG_GEOGRAPHY_EEA)
            "NOT_EEA", "NON_EEA" -> builder.setDebugGeography(ConsentDebugSettings.DebugGeography.DEBUG_GEOGRAPHY_NOT_EEA)
            else -> { /* leave default geography */ }
        }
        devices.forEach { builder.addTestDeviceHashedId(it) }

        return builder.build()
    }

    /**
     * Maps ConsentInformation.consentStatus (int) to the PHP string constants in
     * BlessedZulu\NativePhpAdmob\Events\ConsentChanged.
     */
    fun statusString(info: ConsentInformation): String = when (info.consentStatus) {
        ConsentInformation.ConsentStatus.NOT_REQUIRED -> "not_required"
        ConsentInformation.ConsentStatus.REQUIRED -> "required"
        ConsentInformation.ConsentStatus.OBTAINED -> "obtained"
        else -> "unknown"
    }

    fun isFormRequired(info: ConsentInformation): Boolean =
        info.consentStatus == ConsentInformation.ConsentStatus.REQUIRED
}
